### DISCUSSION TITLE
Is Moor support for Multi-thread?

### DISCUSSION THREAD

**Discussion by rrifafauzikomara** [simolus3/drift#1187] [Q&A]
Is the `Moor` support for `Multi-thread` like in `SQLite`?

> Multi-thread. In this mode, SQLite can be safely used by multiple threads provided that no single database connection is used simultaneously in two or more threads.

References: [https://sqlite.org/threadsafe.html#:~:text=Multi%2Dthread.,in%20two%20or%20more%20threads](https://sqlite.org/threadsafe.html#:~:text=Multi%2Dthread.,in%20two%20or%20more%20threads)

**Comment by simolus3** [comment]
It depends on what exactly you're doing across threads. In general, you can access the same database file across multiple isolates. When sqlite3 is compiled with thread-safety enabled, queries will have a consistent view across threads as you'd expect. However, some moor features that are built on top of sqlite3 apis require some setup to work across threads, mainly stream queries and migrations.

For stream queries, you need to use a synchronization mechanism so that different isolates tell each other about updates they're making. This can be done with the [isolates](https://moor.simonbinder.eu/docs/advanced-features/isolates/) library provided by moor. That library makes it easy to have a single (background) isolate opening the database. Other isolates would then talk to the background isolate to send updates and synchronize streams. Apart from the slightly complicated setup, moor completely hides the fact that the database is running on another isolate.
If you don't want to use stream queries, you can open the same database on multiple isolates without them talking to each other. However, you really need to make sure that you're not invoking the migration concurrently, __as it is not thread safe__ by default. So you need some kind of synchronization mechanism to ensure that the same database is not opened by multiple isolates as the same time. If you use the moor isolate library, there's only a single isolate so this is not a problem.
A common way to use multi-threaded databases without worrying about concurrency with opening or stream-query synchronization is to just use the [MultiExecutor](https://pub.dev/documentation/moor/latest/moor/MultiExecutor-class.html) class. You can use a regular `MoorIsolate` for writes on a background thread while running reads on the main isolate. This avoids doing too much work on the main isolate while also working around the increased latency of cross-isolate communication. That mode of operation works especially well when [WAL](https://sqlite.org/wal.html) is enabled.
